Add configurable filters to mirrors

psakievich opened the pull request:
Add the ability to filter what specs get pushed to a mirror. This allows for a more controlled autopush for binary mirrors. 

Currently filtering can only be called on `spack mirror create` and `spack mirror create` will only create a source mirror and NOT a binary mirror. Refactoring all of that I considered out of scope for this effort since that would be a refactor of most of the module and need a rethink on the user API.

Since source mirror creation also does population the filtering only gets preserved during the lifetime of the method.

For binary mirrors we need the filters to last during a `spack install` for `--autopush` so the filter specs need to be preserved. The implementation in this PR is to recorded the filter specs for binary mirrors in the `mirrors.yaml` because I am considering them a property of the mirror.

Requested to not merge until after #48713

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kwryankrattiger commented on lib/spack/spack/mirrors/filter.py:
```diff
@@ -0,0 +1,38 @@
+        skip  | keep  | outcome
+        ------------------------
+        False | False | Keep
+        True  | True  | Keep
```
I think this is okay, but I think calling the second field include is kind of strange since things that don't match the include are also included. Adding a `@:` match for include disables the exclude, which makes the meaning of exclude kind of wonky.

Really this is a exclude, and override/exempt from exclude, which is a little different than include. Maybe for each exclusion there could be an optional exemptions field?

```yaml
exclude:
    - match: dev_path=*
      exempt:
          - zlib
          - ~shared
    - intel-oneapi-runtime
```

haampie replied:
just be consistent with views/modules etc. https://spack.readthedocs.io/en/latest/environments.html#advanced-view-configuration

kwryankrattiger replied:
The main issue with that is to get detailed exclusions in a dev environment, where the majority of dev specs should be excluded, is you need to use `<package_name> dev_path=*` to say you want to push one spec to a shared developer build cache. That is potentially a lot of config if there are 10s of dev specs in the environment and leads to more potential for error.

Granted, a tool like `spack-manager` could maintain a reference block in `mirrors.yaml` that managed the full list of develop specs maybe that could work...

```yaml
.spack-manager:
  develop-specs: &sm_mirror_exclusions
    - zlib dev_path=*
    - ncurses dev_path=*

mirrors:
  my_mirror:
     select: "@:"
     exclude:
     - *sm_mirror_exclusions
     - intel-oneapi-runtime
```

haampie replied:
same problem for modules and views...

haampie replied:
https://spack.readthedocs.io/en/latest/module_file_support.html#exclude-or-include-specific-module-files looks like views and modules are already inconsistent and the module way is what you want?

psakievich replied:
@kwryankrattiger I added the `--exclude-from-file` to `spack mirror add` so complex exclusion configurations can be supported. Oh but I suppose what you want is `--include-from-file`.  I can add that.

@haampie  I __think__ what I have matches modules. @kwryankrattiger I agree exemption is likely a better name. I can make that change here, but it is adding another bifurcation of syntax.

psakievich replied:
Overall I would prefer to not add a more advanced syntax here. I think between spec matrices, views and modules we should refine and consolidate, then be consistent here. Straight lists are a neutral party between all these cases.

kwryankrattiger replied:
I think following existing configs is a better course than inventing something new. The file support is fine to handle these more complex cases, the perennial problem then becomes more tooling to be able to generate those files.

psakievich commented:
Converting to draft until #48713 merges

psakievich commented:
@haampie @kwryankrattiger bump

psakievich committed:
Add configurable filters to mirrors

Add the ability to filter what specs get pushed to a mirror.
This allows for a more controlled autopush for binary mirrors.

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich commented:
Mega squash to satisfy DSO

kwryankrattiger commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -428,6 +491,8 @@ def _configure_mirror(args):
         changes["source"] = "source" in args.type
 
     changed = entry.update(changes, direction)
+    if hasattr(args, "include_file"):
```
I don't see why this check is needed.

psakievich replied:
The method is called to handle parsing for `set-url` and `set`. It is only valid in `set`. It is ugly but I didn't want to refactor this whole file as part of the PR since it was pending on #48713.

kwryankrattiger replied:
ah I see now, the diff in github made it look liek this was in the `mirror_add`  function at first glance.

kwryankrattiger commented on lib/spack/docs/mirrors.rst:
```diff
@@ -120,6 +120,16 @@ what it looks like:
 
+Customization of the mirror contents can be done by selectively excluding
+specs using the ``--exclude-file`` or ``--exclude-specs`` flags with
+``spack mirror create``.  Note that these only apply to source mirrors.
```
```suggestion
``spack mirror [add|set]``.
```

These options are added for `spack mirror add` and `spack mirror set` but not `spack mirror create`.

psakievich replied:
`spack mirror create` populates source files when it runs so the two flags in the docs will only apply to source files. I suppose we could change that to these flags apply uniformly to source and binary, but when you run `spack mirror create` it will populate a source mirror and not a binary mirror.

I was attempting to document our (IMHO poor) API.

kwryankrattiger replied:
My point here is you are adding these arguments to add/set but the doc update here doesn't reflect that.

psakievich replied:
I'll update the docs to see if I can explain it all better. Good catch. I'm surprised I didn't add the new flags with their commands.

kwryankrattiger commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -299,6 +335,30 @@ def _default_variable(id_):
         return None
 
 
+def _manage_filters(args, mirror) -> bool:
```
Nit on naming, I think maybe `_collect_mirror_filters` or `_update_mirror_filters` may be better.

And pass mirror first (since that is what is being operated on) and then args.

kwryankrattiger commented on lib/spack/spack/mirrors/mirror.py:
```diff
@@ -97,6 +97,11 @@ def display(self, max_len=0):
         binary = "b" if self.binary else " "
         print(f"{self.name: <{max_len}} [{source}{binary}] {url}")
 
+    def _process_spec_filters(self, key: str) -> List[str]:
```
This should follow the naming convention we use for the other fields 

```suggestion
    def _get_spec_filters(self, key: str) -> List[str]:
```

kwryankrattiger commented on lib/spack/spack/schema/mirrors.py:
```diff
@@ -77,6 +77,8 @@
         "fetch": fetch_and_push,
         "push": fetch_and_push,
         "autopush": {"type": "boolean"},
+        "exclude": {"type": "array", "items": {"type": "string"}},
```
should these also support the file or list?

```
exclude: spec_exclusion_file.txt
```
or
```
exclude:
- intel-oneapi-mkl
- ...
```

psakievich replied:
I made it so the file contents get added to the list in the yaml. I can change that though if we want to keep the file as the source of truth.

psakievich replied:
I did it this way to not have to support a merge for a file and yaml, but I suppose the mirror definition would be one or the other this way so it would work.

kwryankrattiger replied:
yeah I think it would be nice, this list can get quite long so not having it in the YAML is an option I am looking for.

psakievich replied:
Hmm I'm wondering if I should make the list a spec or file? Then you could have multiple files and/or a combination rather than list of specs or single file.  Thinking about parsing this

```
spack mirror set foo --exclude-specs a b c --exclude-file blah.txt
```

Right now this is valid. I suppose I can make them mutually exclusive, but I don't see why they have to be.

psakievich replied:
Eh NVM. I  updated to preserve the file and added mutual exclusivity. I think simple is best for now. It shouldn't be too hard to add additional flexibility, but I also don't think we need to over engineer this right now.

kwryankrattiger reviewed (commented):
Took another pass on the implementation, still looking at the tests.

psakievich commented:
@kwryankrattiger  I updated the description with some additional context on the feature's implementation and scope. I would argue our mirror command is confusing and bifurcates with `buildcache` too much. But refactoring all of that for this feature seems like massive scope creep to me.

psakievich committed:
Add ability to store filter file in config

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Update docs

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Try to captures system exit

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Make unit-test finally pass

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Update parsing to preserve file in the config

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Differentiate skip vs filter

Make sure message from a buildcache push indicates why it was not there.

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Merge branch 'develop' into psakiev/cache-no-dev

Signed-off-by: Phil Sakievich <psakiev@sandia.gov>

psakievich committed:
style

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Style again

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Fix missed test

Signed-off-by: psakiev <psakiev@sandia.gov>

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -460,13 +460,19 @@ def push_fn(args):
-                "    Use --force to overwrite them.".format(
-                    len(skipped), ", ".join(elide_list([_format_spec(s) for s in skipped], 5))
-                )
+                f"The following {len(skipped)} specs were skipped as they already exist in the"
```
```suggestion
                f"The following {llnl.util.string.plural(len(skipped), "spec")} were skipped as they already exist in the"
```
This will properly display "spec" if it's only one and "specs" for multiple.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -460,13 +460,19 @@ def push_fn(args):
 
+    if excluded:
+        tty.info(
+            f"The following {len(excluded)} specs were ommitted due from pushing to the"
```
Same here for plural

becker33 commented on lib/spack/spack/mirrors/filter.py:
```diff
@@ -0,0 +1,38 @@
+        False | True  | Keep
+        True  | False | Skip
+        """
+        filter = []
+        filtrate = []
```
This may be easier to read with `included` and `excluded` instead of `filter` and `filtrate`, respectively.

psakievich replied:
Point taken. I didn't use those terms to avoid overloading since the filter can be used for inclusion/exclusion. I can change the names if the context is sufficiently clear.

becker33 commented on lib/spack/docs/mirrors.rst:
```diff
@@ -137,6 +137,19 @@ command:
+specs using the ``--exclude-file`` or ``--exclude-specs`` flags with
+``spack mirror create``.
+
+Note that these only apply to source mirrors when using ``spack mirror create``. 
+This is because ``spack mirror create`` only generates source mirrors.
```
I'm not sure I understand what point this is making that is not made in the paragraphs on either side.

psakievich replied:
True, I might be over-emphasizing the point. Personally, I find the lack of symmetry between create/add plus the interplay of the buildcache command to be very confusing.

becker33 commented on lib/spack/docs/mirrors.rst:
```diff
@@ -137,6 +137,19 @@ command:
+Note that these only apply to source mirrors when using ``spack mirror create``. 
+This is because ``spack mirror create`` only generates source mirrors.
+
+The ``--exclude-file`` and ``--exclude-specs`` flags may also be used with ``spack mirror [create | add]``. These will populate an ``exclude`` section in the mirrors configuration.
```
I think this sentence is meant to only discuss the `spack mirror add` command, since the `create` command is already covered and doesn't save the options to config. Am I missing something?

psakievich replied:
That's a good point. I tried to make the flags mirror `create` for common look and feel, but the side-effects of modifying the config is different.

becker33 commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -134,6 +134,7 @@ def setup_parser(subparser: argparse.ArgumentParser) -> None:
         default=None,
         dest="signed",
     )
+    arguments.add_filter_args(add_parser)
```
I think there should be analogous arguments on the `spack buildcache push` command, for the case that you want to restrict an individual command rather than restricting the entire mirror (analogous to the new `spack mirror create` arguments).

becker33 commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -299,6 +301,30 @@ def _default_variable(id_):
         return None
 
 
+def _collect_mirror_filters(mirror, args) -> bool:
```
Can you rename this to `_parse_mirror_filters` or something similar? `_collect` to me implies that we're returning a collection, which isn't true in this case.

becker33 commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -470,7 +501,10 @@ def specs_from_text_file(filename, concretize=False):
         specs_in_file = f.readlines()
         specs_in_file = [s.strip() for s in specs_in_file]
-    return spack.cmd.parse_specs(" ".join(specs_in_file), concretize=concretize)
+    if concretize:
+        return spack.cmd.parse_specs(" ".join(specs_in_file), concretize=True)
+    else:
+        return spack.cmd.parse_specs(specs_in_file, concretize=False)
```
Why the different argument handling for these two cases?

psakievich replied:
IIRC the parser when concretizing input specs needs you to rejoin the specs for the solve.

becker33 commented on lib/spack/spack/cmd/common/arguments.py:
```diff
@@ -632,6 +632,29 @@ def add_concretizer_args(subparser):
     )
 
 
+def add_filter_args(subparser):
```
It would be nice if these didn't need to be exclusive groups (so someone could e.g. exclude a file for a common set of things they exclude, but also exclude an additional spec they need to exclude this time). It looks like that would require minor changes to the config format (an `exclude` block, with separate keys for `specs` and `files`, instead of a single entry under `exclude`) but otherwise seems feasible. Do you think that is workable?

psakievich replied:
Sigh. This is a fair point. I just didn't want to reimplement this extension. The logic for interleaving the specs from different sources plus include and exclude was more complicated to implement than it seems.

Every other request in this review is pretty straightforward. This one I think will take more work. I can take another pass at this. Maybe using an LLM it will go super fast since this is 90% done already. If not can we define an intermediate state and expand this later? My main concern is it is changing config which we want to not change too often.

psakievich replied:
@thomas-bouvier want to weigh in on this? Is the change requested a significant impact to your use-cases?

thomas-bouvier replied:
I don't have a strong opinion on this. Maybe this feature could be a follow-up PR.

becker33 reviewed (changes requested):
Did not review the tests in detail -- generally, the tests look reasonable but they would need to change substantially for the other changes I've requested to cover new corner cases, so I left them out of this review.

thomas-bouvier commented:
What is the status of this PR @psakievich ?

thomas-bouvier commented:
Would it make sense to filter out develop dependencies by default? This would avoid polluting binary caches

psakievich committed:
Merge pull request #357 from psakievich/claude/cache-no-dev-branch-rtkn3 (#52495)

Address review comments on PR #50099 (configurable mirror filters)

This has been sitting on my fork for a while. Portions of this were
generated by an LLM, but reviewed/owned by me.

Signed-off-by: psakievich <psakievich@gmail.com>
